Error propagation analysis due to fixed point arithmetic for multiplication and squaring

tanujkhattar opened the pull request:
@ncrubin This is a first attempt to reproduce the error analysis for fixed point arithmetic for multiplication and squaring a real number from https://arxiv.org/pdf/2007.07391.pdf

Ideally, the bloqs corresponding to these primitives would expect registers of `QFxp` data type and running classical simulation on the corresponding composite bloq should reproduce this analysis. But the quantum data types are still being implemented so this notebook serves as a reference and a good starting point to reproduce the analysis from the paper. 

@fdmalone  @mpharrigan  - This would add an additional dependency to fxpmath, which is MIT licensed.

mpharrigan commented on dev_tools/requirements/envs/dev.env.txt:
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with Python 3.10
+# This file is autogenerated by pip-compile with Python 3.11
```
please use the docker image to compile the requirements

tanujkhattar replied:
done

mpharrigan commented on qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb:
```diff
@@ -0,0 +1,203 @@
+    "\n",
+    "# Initial setup - Helpers\n",
+    "def fxp(x: float, d: int) -> Fxp:\n",
+    "    dtype = f'fxp-u{d}/{d}'\n",
```
docstring is necessary here, esp if one is not familiar with fxpmath

tanujkhattar replied:
added docstrings

mpharrigan commented on qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb:
```diff
@@ -0,0 +1,203 @@
+    "    except:\n",
+    "        print(f'Failed! {d=}, {eps=}')\n",
+    "\n",
+    "for eps in [1e-3, 1e-4, 1e-5, 1e-6, 1e-7]:\n",
```
these all succeed

tanujkhattar replied:
Yeah, they are all expected to succeed. They succeed for `d-1` as well, which is interesting. Similarly for the next section the tests actually fail for the `d` value proposed in the paper and need a `+1` for small bitsizes. 

The goal here is to provide a framework to replicate the analysis in the paper.

mpharrigan commented on qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb:
```diff
@@ -0,0 +1,203 @@
+    "        for _ in range(100):\n",
+    "            a, b = rng.random(2)\n",
+    "            res = mul_via_repeated_add(a, b, d)\n",
+    "            assert_allclose(res, a * b, eps)\n",
+    "        print(f'Success! {d=}, {eps=}')\n",
+    "    except:\n",
```
bare except. Is this the best pattern here? asserting and catching?

tanujkhattar replied:
changed to catch only `AssertionError` instead of bare except

mpharrigan commented on qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb:
```diff
@@ -0,0 +1,203 @@
+   "source": [
+    "# Multiplying two real numbers\n",
+    "def get_bitsize_for_fxp_mul(eps: float):\n",
+    "    return int(np.ceil(1 + np.log2(1/eps)) + np.log2(1 + np.log2(1/eps)))\n",
```
this is from the paper?

tanujkhattar replied:
Yes, I'll add a comment to reference equation.

mpharrigan reviewed (commented):
this is just numerically testing the error formulas from the appendices, yes? If that is the full scope, I'd suggest changing the title of the notebook.

The code in the notebook needs a lot of documentation

mpharrigan commented:
Taking these implementations and using them for the `on_classical_vals` overrides would be very cool 👍

tanujkhattar commented:
> this is just numerically testing the error formulas from the appendices, yes? 

Yup, but the eventual goal of this notebook is to demonstrate this analysis by instantiating appropriate bloqs. This is a first iteration that gets the analysis in place and follow up PRs would add it to appropriate bloqs and update the notebook with bloq instantiation and classical simulation 

> If that is the full scope, I'd suggest changing the title of the notebook.
Do you have suggestions? Maybe "Numerical analysis of error propagation due to fixed point arithmetic" ?

fdmalone commented:
This is a bit off topic but do we have a plan in place for updating classical simulation to allow for non uint64 types?

More on topic another useful primitive to add would be [scaling an float by an integer ](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/arithmetic/multiplication.py#L242) (should be renamed) which has some interesting shifting / precision properties I hadn't really thought about before.

fdmalone commented on qualtran/bloqs/arithmetic/Numerical analysis of error propagation due to fixed point arithmetic.ipynb:
```diff
@@ -0,0 +1,189 @@
+{
```
Are we ok with spaces in filenames?

tanujkhattar replied:
Renamed to remove spaces and match pattern used by other notebooks

tanujkhattar commented:
> More on topic another useful primitive to add to your notebook would be [scaling a real by an integer ](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/arithmetic/multiplication.py#L242)(should be renamed) which has some interesting shifting / precision properties I hadn't really thought about before

@fdmalone I've added a section to the notebook that analyzes the scaling a real by an integer primitive following Appendix D4 of the paper. 

I think this is good to go now and we can revisit once we have updated classical simulation code to support `Fxp` along with data types.

fdmalone reviewed (approved):
LGTM
